CASMINST-3902: Add and update procedures related to adding river cabinets to a system (#2344)
Cray-HPE/docs-csm: operations/security_and_authentication/Configure_root_user_on_HPE_iLO_BMCs.md

# Configure root user on HPE iLO BMCs

By default, HPE BMC controllers have the `Administrator` user account. In order to discover this type of hardware, the `root` service account needs to be configured.

> When is this procedure applicable:

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR5

> 1. Root password is known & does not match the destination system default air-cooled BMC credentials. For example, a node has been moved between systems and each system gas different default global credentials.

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR7

> 2. Root user does not exist.

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR9

> 3. Root user exists with an unknown password.

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR10

## Prerequisites

- The BMC is accessible over the network via hostname or IP address.

## Procedure

> 1. (`ncn#`) Retrieve the root user password for this BMC.

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR18

>     1. **If configuring a BMC already present in the system**, retrieve the the device specific root user password from Vault:

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR20

        ```bash
        BMC=x3000c0s3b0
        EXPECTED_ROOT_PASSWORD=$(cray scsd bmc creds list --targets $BMC --format json | jq .Targets[].Password -r)
        ```

>         If the following output indicates that Vault does not contain a device specific root user password for the specified BMC. Please use the system default air-cooled BMC root password described in the step below.

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR27

        ```text
        jq: error (at <stdin>:3): Cannot iterate over null (null)
        ```

>     1. **If configuring a new BMC being added to the system**, retrieve the system's default air-cooled BMC root user password from Vault:

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR33

        ```bash
        VAULT_PASSWD=$(kubectl -n vault get secrets cray-vault-unseal-keys -o json |
                            jq -r '.data["vault-root"]' |  base64 -d)
        EXPECTED_ROOT_PASSWORD=$(kubectl -n vault exec -it cray-vault-0 -c vault -- env \

>             VAULT_TOKEN=$VAULT_PASSWD VAULT_ADDR=http://127.0.0.1:8200 VAULT_FORMAT=json \

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR39

            vault kv get secret/reds-creds/defaults | jq .data.Cray.password -r)
        ```

> 1. (`ncn#`) Verify the contents of `EXPECTED_ROOT_PASSWORD`:

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR43

    ```bash
    echo $EXPECTED_ROOT_PASSWORD
    ```

> 1. (`ncn#`) Set an environment variable containing the hostname or current IP Address of the BMC. If coming from the [Add Worker, Storage or Master NCNs](../node_management/Add_Remove_Replace_NCNs.md#add-worker-storage-master)

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR49

>     procedure, then the IP address of should stored in the `BMC_IP` environment variable.

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR50

    Via hostname:

    ```bash
    export BMC=x3000c0s3b0
    ```

    Via IP address:

    ```bash
    export BMC=10.254.1.9

>     ```

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fL51-L62

> 1. (`ncn#`) Determine if the root user account is functional:

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR64

    ```bash
    curl -k -u "root:$EXPECTED_ROOT_PASSWORD" https://$BMC/redfish/v1/Managers -i  | head -1
    ```

    Expected output of a functional root user account, **if this is observed no further action is required to enable the root user account. The rest of the procedure can be skipped**.

    ```text
    HTTP/1.1 200 OK

>     ```

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR70-R76

>     If the following is observed this indicates the root user either does not exist or is configured with a different password. **Continue this procedure to correct the root user credentials.**

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR78

    ```text
    HTTP/1.1 401 Unauthorized
    ```

1. (`ncn#`) Determine the default user credentials:

    > `read -s` is used to read the password in order to prevent it from being echoed to the screen or saved in the shell history.
    > Note that the subsequent `curl` commands **will** do both of these things. If this is not desired, the call should be made in
    > another way.

>     1. The default user name is `Administrator`. Default credentials for the Administrator user on HPE nodes can be found on the serial label pull out tab attached to the server. See [this page for more information](https://support.hpe.com/hpesc/public/docDisplay?docId=sf000046874en_us&docLocale=en_US).

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR90

        ```bash
        export DEFAULT_USERNAME=Administrator
        read -s DEFAULT_PASSWORD

>         export DEFAULT_PASSWORD

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fL90-L93

        ```

>     1. Verify contents of DEFAULT_USERNAME and DEFAULT_PASSWORD:

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/c159690e8aac724540253a1aa495217b2507775d

        ```bash
        echo $DEFAULT_USERNAME

>         echo $DEFAULT_PASSWORD

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fL99

        ```

> 1. (`ncn#`) Try to access the BMC with the default user credentials:

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR103

    ```bash
    curl -k -u "$DEFAULT_USERNAME:$DEFAULT_PASSWORD" https://$BMC/redfish/v1/Managers -i | head -1
    ```

    If a `200 OK` status code is returned, then the default user account is configured correctly.

    ```text
    HTTP/1.1 200 OK
    ```

>     If a `401 Unauthorized` status code is returned, the the default user is not configured correctly. The BMC needs to be factory reset to restore the default user credentials.

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR115

    ```text
    HTTP/1.1 401 Unauthorized
    ```

> 1. (`ncn#`) **Do only if a `401 Unauthorized` was observed in the previous step**, factory reset BMCs back to factory defaults:

Reviewer: ```suggestion
1. (`ncn-mw#`) **Only if a `401 Unauthorized` was observed in the previous step**, then reset BMCs back to factory defaults.
```

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR121

>     1. Follow [this HPE support article](https://techlibrary.hpe.com/docs/iss/proliant-gen10-uefi/s_reset_ilo_defaults.html) to boot the node into its BIOS and perform a factory reset on the BMC.

Reviewer: ```suggestion
    1. Follow [this HPE support article](https://techlibrary.hpe.com/docs/iss/proliant-gen10-uefi/s_reset_ilo_defaults.html) to boot the node into its BIOS and perform a factory reset on the BMC.

```

>         > **NOTE**: If this is a Apollo 6500 XL 645D Gen10 Plus node the BMC will need to be reconfigured with the expected network settings.

Reviewer: ```suggestion
        > **NOTE**: If this is a Apollo 6500 XL 645D Gen10 Plus node, then the BMC will need to be reconfigured with the expected network settings.
```

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR125

        > See [Configure HPE Apollo 6500 XL645d Gen10 Plus Compute Nodes](../../install/prepare_compute_nodes.md#configure-hpe-apollo-6500-xl645d-gen10-plus-compute-nodes) for more information.

>     1. Verify the the BMC can be accessed with the default credentials after the factory reset has been performed.

Reviewer: ```suggestion
    1. Verify that the BMC can be accessed with the default credentials after the factory reset has been performed.
```

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR128

        ```bash
        curl -k -u "$DEFAULT_USERNAME:$DEFAULT_PASSWORD" https://$BMC/redfish/v1/Managers -i  | head -1
        ```

        Expected output:

        ```text
        HTTP/1.1 200 OK
        ```

> 1. (`ncn#`) Configure the root user account:

Reviewer: ```suggestion
1. (`ncn-mw#`) Configure the root user account.
```

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR140

>     1. Determine if the root user account already exists:

Reviewer: ```suggestion
    1. Determine if the root user account already exists.
```

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR142

        ```bash

>         for account in $(curl -s -u $DEFAULT_USERNAME:$DEFAULT_PASSWORD -k https://$BMC/redfish/v1/AccountService/Accounts | jq '.Members[]."@odata.id"' -r); do 
>             echo "Checking $account"; 
> 
>             curl -k -s -u $DEFAULT_USERNAME:$DEFAULT_PASSWORD -k https://${BMC}${account} | jq '. | {Id: .Id, UserName: .UserName, RoleId: .RoleId}' -c

Reviewer: ```suggestion
        for account in $(curl -s -u "$DEFAULT_USERNAME:$DEFAULT_PASSWORD" -k https://${BMC}/redfish/v1/AccountService/Accounts | jq '.Members[]."@odata.id"' -r); do 
            echo "Checking $account"
            curl -k -s -u "$DEFAULT_USERNAME:$DEFAULT_PASSWORD" -k https://${BMC}${account} | jq '. | {Id: .Id, UserName: .UserName, RoleId: .RoleId}' -c
```

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR145-R148

        done
        ```

        Example output if the root user exists:

>         ```json

Reviewer: ```suggestion
        ```text
```

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR162

        Checking /redfish/v1/AccountService/Accounts/1
        {"Id":"1","UserName":"Administrator","RoleId":"Administrator"}
        Checking /redfish/v1/AccountService/Accounts/2
        {"Id":"2","UserName":"root","RoleId":"Administrator"}
        ```

        Example output if the root user does not exist:

>         ```json

Reviewer: ```suggestion
        ```text
```

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR153

        Checking /redfish/v1/AccountService/Accounts/1
        {"Id":"1","UserName":"Administrator","RoleId":"Administrator"}
        ```

>     1. **If the user does not exist**, then create the root user account:

Reviewer: ```suggestion
    1. **If the user does not exist**, then create the root user account.
```

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR167

        ```bash
        curl -k -u "$DEFAULT_USERNAME:$DEFAULT_PASSWORD" -X POST \
                -H 'Content-Type: application/json' \

>                 -d $(jq --arg PASSWORD "$EXPECTED_ROOT_PASSWORD" -nc '{RoleId: "Administrator", UserName: "root", Password: $PASSWORD}') \

Reviewer: Still not good to pass passwords on the command line, but I had never considered using jq this way for creating payloads for curl. Very slick.

Author: Might be able to make the JQ access the password from the environment. https://stedolan.github.io/jq/manual/#$ENV,env

Also in the future I hope this entirely procedure here would be superseded by FR5 in https://github.com/Cray-HPE/community/issues/37

          "Id": "3",
          "Description": "iLO User Account",
          "Links": {
            "Role": {
              "@odata.id": "/redfish/v1/AccountService/Roles/Administrator"
            }
          },
          "Name": "User Account",
          "Oem": {
            "Hpe": {
              "@odata.context": "/redfish/v1/$metadata#HpeiLOAccount.HpeiLOAccount",
              "@odata.type": "#HpeiLOAccount.v2_2_0.HpeiLOAccount",
              "LoginName": "root",
              "Privileges": {
                "HostBIOSConfigPriv": true,
                "HostNICConfigPriv": true,
                "HostStorageConfigPriv": true,
                "LoginPriv": true,
                "RemoteConsolePriv": true,
                "SystemRecoveryConfigPriv": false,
                "UserConfigPriv": true,
                "VirtualMediaPriv": true,
                "VirtualPowerAndResetPriv": true,
                "iLOConfigPriv": true
              },
              "ServiceAccount": false
            }
          },
          "Password": null,
          "PasswordChangeRequired": false,
          "RoleId": "Administrator",
          "UserName": "root"
        }
        ```

>     1. **If the user exists**, then change the root password:

Reviewer: ```suggestion
    1. **If the user exists**, then change the root password.
```

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR219

>         1. Determine the ID associated with the root account. In the example output above the root user ID is `2`.

Reviewer: ```suggestion
        1. Determine the ID associated with the root account. In the example output above, the root user ID is `2`.
```

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR221

             ```bash

>              export ROOT_USER_ACCOUNT_ID=2

Reviewer: ```suggestion
             ROOT_USER_ACCOUNT_ID=2
```

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR224

             ```

>         1. Using the default administrator credentials change the root account password:

Reviewer: ```suggestion
        1. Using the default administrator credentials, change the root account password.
```

Author: https://github.com/Cray-HPE/docs-csm/pull/2344/commits/ab3c45572fa5b86e6e61bbd196c6734919b6dae8#diff-330e293e8b0ca943ab16c00a167855bd2869e2a91837546c10d2d26b0cd16d8fR227

            ```bash
            curl -k -u "$DEFAULT_USERNAME:$DEFAULT_PASSWORD" -X PATCH \
                -H 'Content-Type: application/json' \
                -d $(jq --arg PASSWORD "$EXPECTED_ROOT_PASSWORD" -nc '{Password: $PASSWORD}') \
                https://$BMC/redfish/v1/AccountService/Accounts/$ROOT_USER_ACCOUNT_ID | jq
            ```

            Expected output:

            ```json
            {
              "error": {
                "code": "iLO.0.10.ExtendedInfo",
                "message": "See @Message.ExtendedInfo for more information.",
                "@Message.ExtendedInfo": [
                  {
                    "MessageId": "Base.1.4.AccountModified"
                  }
                ]
              }
            }
            ```

> 1. (`ncn#`) Confirm the new credentials can be used with Redfish:

Reviewer: ```suggestion
1. (`ncn-mw#`) Confirm that the new credentials can be used with Redfish.
```
